### PR TITLE
Ability to define RasterSource as a CSV of scene URLs

### PR DESCRIPTION
**Pull request by metasim**
Depends on #144 and #143

### CODE REVIEW

File: pyrasterframes/src/main/python/docs/raster-io.pymd [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+```python
+csv = f"""
+red,green,blue
+{s2_band(4)},{s2_band(3)},{s2_band(2)}
+"""
+
+print(csv)
+
+rf2 = spark.read.rastersource(csv=csv)
```
Comment (metasim) [inline pyrasterframes/src/main/python/docs/raster-io.pymd]: @vpipkt Thought you might enjoy this...

**Comment by metasim** [comment]
> 1. Support for `list` of `str` (URI's)  into paths param (already have?)

Definitely supported in Scala; need to create test and define

> 2. Support for pandas DF into catalog param

Let's create a separate ticket for this... My initial response is negative, but I need to consider what  biases I'm bringing to the table. Definitely want to make sure we have the API correct for Spark DataFrames before adding that.

> 3. Retain columns other than `catalog_col_names` in the reader (in other words taking the other columns as `RasterMetadata` in the parlance of our [release plan](https://github.com/locationtech/rasterframes/wiki/0.8.0-Release-Planning#datasources); this likely rises to a separate issue.

Good point. Let's make this a separate ticket but prioritize it. I'd like to get this merged.

> I can contribute some to examples like spark-ml and local-algebra let's just chat to deconflict

Please do, but make that a separate PR.

File: pyrasterframes/src/main/python/docs/raster-io.pymd [hunk 2]
```diff
@@ -1,7 +1,155 @@
...
+from IPython.display import display
+from pyrasterframes.utils import create_rf_spark_session 
+from pyrasterframes.rasterfunctions import *
+spark = create_rf_spark_session()
+```
+
+RasterFrames registers a DataSource named `rastersource` that enables reading of GeoTIFFs (and other formats when GDAL is installed) from arbitrary (sets of) URIs. In the examples that follow we'll be reading from a Sentinel 2 scene stored in an AWS S3 bucket:
+
+```python
+data_root = 'https://s22s-test-geotiffs.s3.amazonaws.com/luray_snp'
```
[the remark this answers is not in the document]
Comment (metasim) [reply pyrasterframes/src/main/python/docs/raster-io.pymd]: Yeh, there's no way around that.... you have to use GDAL.

File: pyrasterframes/src/main/python/docs/raster-io.pymd [hunk 3]
```diff
@@ -1,7 +1,155 @@
...
+def s2_band(idx):
+    return f"{data_root}/B0{idx}.tif"
```
[the remark this answers is not in the document]
Comment (metasim) [reply pyrasterframes/src/main/python/docs/raster-io.pymd]: `.format` instead?

[the remark this answers is not in the document]
Comment (metasim) [reply pyrasterframes/src/main/python/docs/raster-io.pymd]: Going through and trying to convert where I see them.

File: pyrasterframes/src/main/python/docs/raster-io.pymd [hunk 4]
```diff
@@ -1,7 +1,155 @@
...
+[[Need to know _a priori_ how many bands there are, specified with `band_indexes`]]
+ 
+
 ## Writing Raster Data
 
+### Overviews
+
+### GeoTIFFs
+
+### Tile Samples
```
[the remark this answers is not in the document]
Comment (metasim) [reply pyrasterframes/src/main/python/docs/raster-io.pymd]: I look to you on what's best with the "user journey"

File: pyrasterframes/src/main/python/pyrasterframes/__init__.py [hunk 5]
```diff
@@ -98,16 +98,46 @@ def _layer_reader(df_reader, format_key, path, **options):
     return _convert_df(df)
 
 
-def _rastersource_reader(df_reader, path=None, band_indexes=None, tile_dimensions=(256, 256), **options):
+def _rastersource_reader(
+        df_reader, path=None,
```
[the remark this answers is not in the document]
Comment (metasim) [reply pyrasterframes/src/main/python/pyrasterframes/__init__.py]: I think there's already a precedent on the Scala-side on how to do this; nevertheless, would like it to be as consistent between modalities as practicable. I believe if it's a newline delimited list of paths it does the union, and the column name is picked for you. If you want a single row with multiple bands, you use a CSV or dataframe.

File: pyrasterframes/src/main/python/docs/raster-io.pymd (hunk 2 above)
[the remark this answers is not in the document]
Comment (vpipkt) [reply pyrasterframes/src/main/python/docs/raster-io.pymd]: We might consider converting the data on the bucket?

File: pyrasterframes/src/main/python/docs/raster-io.pymd [hunk 6]
```diff
@@ -60,10 +60,10 @@ print(tile)
 Generally we'll want to work with multiple bands. If you have several singleband image files of the same scene, `rastersource` can make use of a CSV (or other DataFrame) specifying the band columns and scene rows.
 
 ```python
-csv = f"""
+csv = """
 red,green,blue
-{s2_band(4)},{s2_band(3)},{s2_band(2)}
-"""
+{},{},{}
+""".format(s2_band(4), s2_band(3), s2_band(2))
```
Comment (vpipkt) [inline pyrasterframes/src/main/python/docs/raster-io.pymd]: yikes. maybe as below ?

```python
csv = "red,green,blue\n" + ','.join([s2_band(b) for b in [4, 3, 2]])

```

Comment (metasim) [reply pyrasterframes/src/main/python/docs/raster-io.pymd]: I'm sure you or someone will hack all that away at some point. Just wanted to show a new modality.

**Pull request merged**